packages/intent-filler/src/core/filler.ts

```diff
@@ -0,0 +1,244 @@
+import { chainIds } from "@/config/chain"
+import { EventMonitor } from "./event-monitor"
+import { FillerStrategy } from "@/strategies/base"
+import { Order, FillerConfig, ChainConfig, DUMMY_PRIVATE_KEY, ADDRESS_ZERO, bytes20ToBytes32 } from "hyperbridge-sdk"
+import pQueue from "p-queue"
+import { ChainClientManager, ChainConfigService, ContractInteractionService } from "@/services"
+import { fetchTokenUsdPriceOnchain } from "@/utils"
+import { PublicClient } from "viem"
+
+export class IntentFiller {
+	public monitor: EventMonitor
+	private strategies: FillerStrategy[]
+	private chainQueues: Map<number, pQueue>
+	private globalQueue: pQueue
+	private configService: ChainConfigService
+	private chainClientManager: ChainClientManager
+	private contractService: ContractInteractionService
+	private pendingOrders: Map<string, NodeJS.Timeout> = new Map()
+	private orderRecheckCount: Map<string, number> = new Map()
+	private config: FillerConfig
+
+	constructor(chainConfigs: ChainConfig[], strategies: FillerStrategy[], config: FillerConfig) {
+		this.monitor = new EventMonitor(chainConfigs)
+		this.strategies = strategies
+		this.config = config
+		this.configService = new ChainConfigService()
+		this.chainClientManager = new ChainClientManager(DUMMY_PRIVATE_KEY)
+		this.contractService = new ContractInteractionService(this.chainClientManager, DUMMY_PRIVATE_KEY)
+		this.chainQueues = new Map()
+		chainConfigs.forEach((chainConfig) => {
+			// 1 order per chain at a time due to EVM constraints
+
+			this.chainQueues.set(chainConfig.chainId, new pQueue({ concurrency: 1 }))
+		})
+
+		this.globalQueue = new pQueue({
+			concurrency: config.maxConcurrentOrders || 5,
+		})
+
+		// Set up event handlers
+		this.monitor.on("newOrder", ({ order }) => {
+			this.handleNewOrder(order)
+		})
+	}
+
+	public start(): void {
+		this.monitor.startListening()
+	}
+
+	public stop(): void {
+		this.monitor.stopListening()
+
+		// Clear all pending order timeouts
+		this.pendingOrders.forEach((timeout) => clearTimeout(timeout))
+		this.pendingOrders.clear()
+
+		// Clear recheck counts
+		this.orderRecheckCount.clear()
+
+		// Wait for all queues to complete
+		const promises = []
+		this.chainQueues.forEach((queue) => {
+			promises.push(queue.onIdle())
+		})
+		promises.push(this.globalQueue.onIdle())
+
+		Promise.all(promises).then(() => {
+			console.log("All orders processed, filler stopped")
+		})
+	}
+
+	// Operations
+
+	private handleNewOrder(order: Order): void {
+		// Use the global queue for the initial analysis
+		// This can happen in parallel for PublicClient orders
+		this.globalQueue.add(async () => {
+			try {
+				// Check if the order has enough confirmations
+				const hasEnoughConfirmations = await this.checkConfirmations(order)
```

Comment:
I think it makese sense to instead do a `waitForConfirmations` rather than this setTimeout rechecking mechanism. That would be more granular and less likely to miss when orders reach the required confirmations